dlib/matrix/matrix_fft.h

```diff
@@ -263,6 +264,291 @@ namespace dlib
         }
     }
 
+// ----------------------------------------------------------------------------------------
+
+    namespace details
+    {
+        struct fft_func
+        {
+            template<typename MAT, typename T = typename MAT::type, typename std::enable_if<is_complex<T>::value, bool>::type = true>
+            auto operator()(const MAT& mat) const { return dlib::fft(mat); }
+
+            template<typename MAT, typename T = typename MAT::type, typename std::enable_if<!is_complex<T>::value, bool>::type = true>
+            auto operator()(const MAT& mat) const { return dlib::fft(dlib::complex_matrix(mat)); }
+
+            static constexpr std::size_t freqsize(std::size_t fftsize) { return fftsize; }
+        };
+
+        struct fftr_func
+        {
+            template<typename MAT>
+            auto operator()(const MAT& mat) const { return dlib::fftr(mat); }
+
+            static constexpr std::size_t freqsize(std::size_t fftsize) { return dlib::fftr_nc_size(fftsize); }
+        };
+
+        struct ifft_func
+        {
+            template<typename MAT>
+            auto operator()(const MAT& mat) const { return dlib::ifft(mat); }
+        };
+
+        struct ifftr_func
+        {
+            template<typename MAT>
+            auto operator()(const MAT& mat) const { return dlib::ifftr(mat); }
+        };
+
+        template <
+            typename EXP,
+            typename WINDOW,
+            typename FFT_FUNC
+        >
+        auto stft_impl (
+            const matrix_exp<EXP>& signal,
+            const WINDOW& w,
+            std::size_t fftsize,
+            std::size_t wlen,
+            std::size_t hoplen,
+            const FFT_FUNC& fft_obj
+        )
+        {
+            using T = typename EXP::type;
+            using R = remove_complex_t<T>;
+            using C = add_complex_t<T>;
+
+            static_assert(std::is_floating_point<R>::value, "underlying type must be real or complex floating point type");
+            DLIB_ASSERT(is_vector(signal), "input must be a vector type");
+            DLIB_ASSERT(signal.size() >= (long)wlen, "signal.size() >= wlen not satisfied");
+            DLIB_ASSERT(fftsize >= wlen, "fftsize >= wlen not satisfied");
+            DLIB_ASSERT(wlen >= hoplen, "wlen >= hoplen not satisfied");
+
+            /*! Input is left-padded by wlen/2 and right-padded wlen/2 !*/
+            const std::size_t total_padding = wlen;
+            const std::size_t overlap       = wlen - hoplen;
+            const std::size_t nframes       = (signal.size() + total_padding - overlap) / hoplen;
+            matrix<C> stft = zeros_matrix<C>(nframes,
+                                             FFT_FUNC::freqsize(fftsize));
+            matrix<R> win(1,wlen);
+            for (std::size_t i = 0 ; i < wlen ; ++i)
+                win(0, i) = w(i, wlen);
+
+            /*! TODO: reduce extra buffers, e.g. padded !*/
```

Comment:
```suggestion
            // TODO: reduce extra buffers, e.g. padded
```
Is fine.  Only use the `/*! !*/` style for function and class contracts.  That style is special to some of the documentation generation code and is supposed to be a "hey this is a function/class contract" tag.